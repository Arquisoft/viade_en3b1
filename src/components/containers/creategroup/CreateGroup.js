import React, { useState } from 'react';
import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import Typography from '@material-ui/core/Typography';
import { makeStyles } from '@material-ui/core/styles';
import Container from '@material-ui/core/Container';
import NavBar from '../../ui/main/NavBar.js';
import Card from '@material-ui/core/Card';
import { Grid, TextField, CardContent, Checkbox, Snackbar, IconButton } from '@material-ui/core';
import Footer from '../../ui/main/Footer.js';
import SupervisedUserCircleIcon from '@material-ui/icons/SupervisedUserCircle';
import UserCache from '../../../cache/UserCache.js';
import Group from '../../../entities/Group.js';
import GroupsCache from '../../../cache/GroupsCache.js';
import { uploadGroup } from '../../../handler/GroupHandler.js';
import CloseIcon from '@material-ui/icons/Close';
import MuiAlert from '@material-ui/lab/Alert';

const useStyles = makeStyles((theme) => ({
    root: {
        minWidth: 275,
        margin: theme.spacing(5),
        marginTop: theme.spacing(15)
    },
    paper: {
        marginTop: theme.spacing(5),
        marginBottom: theme.spacing(5),
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
    },
    avatar: {
        margin: theme.spacing(1),
        backgroundColor: '#7c4dff',
    },
    form: {
        width: '80%',
        marginTop: theme.spacing(1),
        alignItems: 'center',
        justifyContent: 'center',
    },
    submit: {
        margin: theme.spacing(1, 0, 2),
    },
    link: {
        margin: theme.spacing(1, 0, 2),
    }
}));

export default function CreateGroup() {
    const classes = useStyles();

    const [name, setName] = useState("");
    const [members, setMembers] = useState([]);
    const friends = UserCache.getFriends();

    const [open, setOpen] = useState(false);
    const [message, setMessage] = useState("");
    const [severity, setSeverity] = useState("");
    const vertical = 'top';
    const horizontal = 'center';

    function submit(e) {
        e.preventDefault();

        let webIdList = [];
        members.forEach((m) => {
            webIdList.push(m.getWebId());
        });

        let group = new Group(name, webIdList);

        GroupsCache.clearGroups();
        GroupsCache.addGroup(group);
        let statusCode = new Promise((resolve) => {
            uploadGroup(group, (response) => resolve(response));
        });

        statusCode.then((code) => {
            checkStatusCode(code);
        });
    };

    const checkStatusCode = (code) => {
        switch (code) {
            case -1: // error
                openNotif("Couldn't create your new group", 'error');
                break;
            case 0: // success
                openNotif("Your group was successfully saved", 'success');
                break;
            default:
                throw new Error('Unknown Success Code ' + code);
        }
    };

    const openNotif = (text, newSeverity) => {
        setMessage(text);
        setSeverity(newSeverity);
        setOpen(true);
    };

    const closeNotif = () => {
        setOpen(false);
    };

    const handleChange = (index) => {
        let memberList = members; // splice(index, 1)
        let friend = friends[index];
        if (memberList.includes(friend)) {
            let indexOfFriend = memberList.indexOf(friend);
            memberList.splice(indexOfFriend, 1);
        } else {
            memberList.push(friend);
        }

        setMembers(memberList);
    };

    return (
        <div>
            <NavBar />

            <Snackbar
                anchorOrigin={{ vertical, horizontal }}
                open={open}
                action={
                    <React.Fragment>
                        <IconButton
                            aria-label="close"
                            color="inherit"
                            onClick={() => closeNotif()}
                        >
                            <CloseIcon />
                        </IconButton>
                    </React.Fragment>
                }
            >
                <Alert onClose={() => closeNotif()} severity={severity}>
                    {message}
                </Alert>
            </Snackbar>

            <Container component="main" maxWidth="sm">
                <Card className={classes.root} elevation={4}>
                    <CssBaseline />
                    <div className={classes.paper}>
                        <SupervisedUserCircleIcon fontSize="large" style={{ fill: '#7c4dff', }} />
                        <Typography component="h1" variant="h5">Want a new group of friends?</Typography>
                        <div className={classes.form}>
                            <form onSubmit={(e) => submit(e)}>
                                <TextField
                                    required
                                    id="name"
                                    name="name"
                                    label="Name"
                                    fullWidth
                                    defaultValue={name}
                                    autoComplete="routename"
                                    onChange={(e) => setName(e.target.value)}
                                />

                                
                                <div style={{ marginTop: '2rem', marginBottom: '2rem' }}>
                                <Typography variant="subtitle1" color="inherit">Select the members of your group:</Typography>
                                    <div style={{ textAlign: "center"}}>
                                    {friends.map((each, index) => (
                                        <Grid key={index} container style={{ textAlign: "center", marginTop: '0.5rem', marginBottom: '0.5rem' }}>
                                            <Checkbox
                                                onChange={() => handleChange(index)}
                                                color="primary"
                                            />
                                            <Card variant="outlined"
                                                style={{ width: '18rem' }}
                                            >
                                                <CardContent>
                                                    <Typography>{each.getName()}</Typography>
                                                </CardContent>
                                            </Card>
                                        </Grid>
                                    ))}
                                    </div>
                                </div>
                                <div style={{ textAlign: 'center' }}>
                                    <Button
                                        data-testid="btn-create-group"
                                        id="btn-create-group"
                                        variant="contained"
                                        color="primary"
                                        className={classes.submit}
                                        type="submit"
                                    >Create</Button>
                                </div>
                            </form>
                        </div>
                    </div>
                </Card>
            </Container>
            <Grid
                item
                xs={12}
                sm={12}
                md={12}
                style={{ marginTop: '36rem' }}
            >
                <Footer />
            </Grid>
        </div>
    );
}

function Alert(props) {
    return <MuiAlert elevation={6} variant="filled" {...props} />;
}