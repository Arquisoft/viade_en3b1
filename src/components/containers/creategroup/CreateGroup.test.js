import CreateGroup from './CreateGroup';

test('Define CreateGroup', () => {
  expect(CreateGroup).toBeDefined();
});

test('Renders correctly', () => {
  const wrapper = render(<CreateGroup />);
  expect(wrapper).toMatchSnapshot();
});