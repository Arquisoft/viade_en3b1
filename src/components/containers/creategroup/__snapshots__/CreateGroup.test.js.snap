// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly 1`] = `
<div>
  <div
    class="makeStyles-root-7"
  >
    <header
      class="MuiPaper-root MuiAppBar-root MuiAppBar-positionStatic MuiAppBar-colorPrimary MuiPaper-elevation4"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular makeStyles-toolbar-8 MuiToolbar-gutters"
      >
        <div
          class="MuiGrid-root makeStyles-link-10"
        >
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <a
              class="MuiTypography-root MuiLink-root MuiLink-underlineNone MuiTypography-colorInherit"
              href="#/home"
            >
              <div
                class="MuiGrid-root MuiGrid-container"
              >
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <img
                    alt="Viade Logo"
                    src="logo_2.svg"
                    style="height:2.5rem;width:2.5rem"
                  />
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item"
                >
                  <h4
                    class="MuiTypography-root MuiTypography-h4"
                  >
                    ViaDe
                  </h4>
                </div>
              </div>
            </a>
          </div>
        </div>
      </div>
    </header>
  </div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthSm"
  >
    <div
      class="MuiPaper-root MuiCard-root makeStyles-root-1 MuiPaper-elevation4 MuiPaper-rounded"
    >
      <div
        class="makeStyles-paper-2"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          style="fill:#7c4dff"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.99 2c-5.52 0-10 4.48-10 10s4.48 10 10 10 10-4.48 10-10-4.48-10-10-10zm3.61 6.34c1.07 0 1.93.86 1.93 1.93 0 1.07-.86 1.93-1.93 1.93-1.07 0-1.93-.86-1.93-1.93-.01-1.07.86-1.93 1.93-1.93zm-6-1.58c1.3 0 2.36 1.06 2.36 2.36 0 1.3-1.06 2.36-2.36 2.36s-2.36-1.06-2.36-2.36c0-1.31 1.05-2.36 2.36-2.36zm0 9.13v3.75c-2.4-.75-4.3-2.6-5.14-4.96 1.05-1.12 3.67-1.69 5.14-1.69.53 0 1.2.08 1.9.22-1.64.87-1.9 2.02-1.9 2.68zM11.99 20c-.27 0-.53-.01-.79-.04v-4.07c0-1.42 2.94-2.13 4.4-2.13 1.07 0 2.92.39 3.84 1.15-1.17 2.97-4.06 5.09-7.45 5.09z"
          />
        </svg>
        <h1
          class="MuiTypography-root MuiTypography-h5"
        >
          Want a new group of friends?
        </h1>
        <div
          class="makeStyles-form-4"
        >
          <form>
            <div
              class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
                data-shrink="false"
                for="name"
                id="name-label"
              >
                Name
                <span
                  class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
                >
                   *
                </span>
              </label>
              <div
                class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
              >
                <input
                  aria-invalid="false"
                  autocomplete="routename"
                  class="MuiInputBase-input MuiInput-input"
                  id="name"
                  name="name"
                  required=""
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              style="margin-top:2rem;margin-bottom:2rem"
            >
              <h6
                class="MuiTypography-root MuiTypography-subtitle1 MuiTypography-colorInherit"
              >
                Select the members of your group:
              </h6>
              <div
                style="text-align:center"
              />
            </div>
            <div
              style="text-align:center"
            >
              <button
                class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-5 MuiButton-containedPrimary"
                data-testid="btn-create-group"
                id="btn-create-group"
                tabindex="0"
                type="submit"
              >
                <span
                  class="MuiButton-label"
                >
                  Create
                </span>
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </main>
  <div
    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-12 MuiGrid-grid-md-12"
    style="margin-top:36rem"
  >
    <div>
      <footer
        class="makeStyles-footer-309"
      >
        <div
          class="MuiContainer-root MuiContainer-maxWidthXs"
        >
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-5"
          >
            <div
              class="MuiGrid-root MuiGrid-item"
            >
              <h6
                class="MuiTypography-root MuiTypography-h6"
              >
                ViaDe En3B1
              </h6>
              <h6
                class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-colorTextSecondary"
              >
                Made by 
                <a
                  class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                  href="https://github.com/PabloCanalSuarez"
                  style="color:#302c58"
                  target="_blank"
                >
                  Pablo Cañal
                </a>
              </h6>
            </div>
            <hr
              class="MuiDivider-root MuiDivider-flexItem MuiDivider-vertical"
            />
            <div
              class="MuiGrid-root MuiGrid-item"
              style="margin-top:0.5rem"
            >
              <a
                class="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorPrimary"
                href="https://github.com/Arquisoft/viade_en3b1/"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 .3a12 12 0 0 0-3.8 23.4c.6.1.8-.3.8-.6v-2c-3.3.7-4-1.6-4-1.6-.6-1.4-1.4-1.8-1.4-1.8-1-.7.1-.7.1-.7 1.2 0 1.9 1.2 1.9 1.2 1 1.8 2.8 1.3 3.5 1 0-.8.4-1.3.7-1.6-2.7-.3-5.5-1.3-5.5-6 0-1.2.5-2.3 1.3-3.1-.2-.4-.6-1.6 0-3.2 0 0 1-.3 3.4 1.2a11.5 11.5 0 0 1 6 0c2.3-1.5 3.3-1.2 3.3-1.2.6 1.6.2 2.8 0 3.2.9.8 1.3 1.9 1.3 3.2 0 4.6-2.8 5.6-5.5 5.9.5.4.9 1 .9 2.2v3.3c0 .3.1.7.8.6A12 12 0 0 0 12 .3"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </footer>
    </div>
  </div>
</div>
`;
